include/ginkgo/core/config/config.hpp

```diff
@@ -0,0 +1,143 @@
+// SPDX-FileCopyrightText: 2017 - 2024 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#ifndef GKO_PUBLIC_CORE_CONFIG_CONFIG_HPP_
+#define GKO_PUBLIC_CORE_CONFIG_CONFIG_HPP_
+
+
+#include <map>
+#include <string>
+#include <unordered_map>
+
+
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/config/type_descriptor.hpp>
+#include <ginkgo/core/solver/solver_base.hpp>
+
+
+namespace gko {
+namespace config {
+
+class registry;
+
+
+class pnode;
+
+
+/**
+ * parse is the main entry point to create an Ginkgo object based on
+ * some file configuration. It reads a configuration stored as a property tree
+ * and creates the desired type.
+ *
+ * General rules for configuration
+ * 1. all parameter and template usage are according to the class directly. It
+ *    has the same behavior as the class like default setting without specifying
+ *    anything. When the class factory parameters allows `with_<key>(value)`,
+ *    the file configuration will allow `"<key>": value`
+ * 2. all key will use snake_case including template. For example, ValueType ->
+ *    value_type
+ * 3. If the value is not bool, integer, or floating point, we will use string
+ *    to represent everything. For example, we will use string to select the
+ *    enum value. `"baseline": "absolute"` will select the absolute baseline in
+ *    ResidualNorm critrion
+ * 4. `"type"` is the new key to select the class without template type. We also
+ *    prepend the namespace except for gko. For example, we use "solver::Cg" for
+ *    Cg solver. Note. the template type is given by the another entry or from
+ *    the type_descriptor.
+ * 5. We have supports the following datatype with postfix to indicate their
+ *    size: int32, int64, float32, float64, complex<float32>, complex<float64>.
+ *    note: we have also allow `void` additionally in type_descriptor to specify
+ *          file must contain the value/index type config.
+ * 6. We use [real, imag] to represent complex values. If it only contains one
+ *    value or none [], we will treat it as a complex number with an imaginary
+ *    part = 0.
```

Comment:
```suggestion
 * 6. Complex values are represented as an 2-element array [real, imag]. If the
 *    array contains only one value, it will be considered as a complex number
 *    with an imaginary part = 0. An empy array will be treated as zero.
```